File: golang/009.go
package main

import "fmt"

func getFirstAndLastOcc(arr []int,target, first, last int)[]int{
	position := binarySearch(arr, target, 0, len(arr))
	positionCopy :=  position
	for arr[position] == target{
		fmt.Println("here")
		first = position
		position --
	}
	for positionCopy < len(arr) && arr[positionCopy] == target{
		last = positionCopy
		positionCopy++
	}
	return []int{first, last}
}

func binarySearch(arr []int, target, start, stop int) int {
	if start > stop{
		return -1
	}

	midVal := start + (stop - start) / 2
	if target == arr[midVal]{
		return midVal
	} else if target > arr[midVal]{
		return binarySearch(arr, target, midVal + 1, stop)
	} else {
		return binarySearch(arr, target, start, midVal - 1)
	}
}

func main(){
	fmt.Println(getFirstAndLastOcc([]int{2,3,4,5,6,7,8,8},8,0,0))
	fmt.Println(binarySearch([]int{2,3,4,5,6,7,8,9},10,0,7))
}

File: python/009.py
def findFirstAndLast(arr, target):
    position = binarySort(arr,target, 0, len(arr)- 1)
    if position == -1:
        return [-1,-1]
    positionCopy = position
    first, last = position
    while arr[position] == target:
        first = position
        position-=1
    while arr[positionCopy] == target:
        last = positionCopy
        positionCopy+=1
    return [first, last]

def binarySort(arr, target, start,stop):
    if start > stop:
        return -1
    
    midVal = start + (stop-start) // 2

    if target == arr[midVal]:
        return midVal

    elif target > arr[midVal]:
        return binarySort(arr, target, midVal + 1, stop)

    else:
        return binarySort(arr, target, start, midVal - 1)
    

print(binarySort([2,3,4,5,6,7,8],8,0,6))
print(findFirstAndLast([3,4,5,5,5,6,7,8],5))
